### ISSUE TITLE
Provide a way to find out how the context menu was closed

### ISSUE THREAD

**Issue by johannest** [vaadin/flow-components#2261]
### Describe your motivation

Currently there is no way to know how the context menu was closed. Some testing tools needs to 'detect and record non-explicit close' (closed by the ESC key or a click outside).

### Describe the solution you'd like

Optimal implementation could be following: There would be an OpenedChangeEvent.getCloseCausedBy() (or 'CloseReason' or 'CloseTrigger') method which returns an enum with the following entries:
- ItemSelected
- OutsideClick
- Programmatically
- Shortcut (ESC Keypress)

Another alternative could be leaving OpenedChangeListener as is and instead introducing new CloseListener which CloseEvent would have the API discussed above.

### Describe alternatives you've considered

Few alternatives which were considered:

1. Try to override the ContextMenu.close menu and use a poor boolean-switch to detect API calls. This does not work in case the menu was closed with an item.
2. Create an add-on (e.g. EnchancedContextMenu) for this purpose. Add-ons are sub optimal because of duplicate code and bad maintenance.

**Comment by yuriy-fix** [comment]
Could you please provide a particular use-case where this API is needed?
Testing tools shouldn't probably be testing the internal logic of closing the `context-menu`.

**Comment by haenni** [comment]
I would like to follow up on this one. 

TL;DR:
We maintain a testing tool to record and replay UI function tests.
In that specific case, we need to know how the ContextMenu was closed so that the replay fires the same event, or we do not record anything (eg: menu was closed programmatically).

Could you please re-evaluate if one of the solutions described in this ticket could be implemented? 
Many thanks.

Our use case in more detail:
 
In our company, we develop ERP software with many different modules and products. While the de-velopers have their unit test suites, we as a company also do testing on the final product. This means, a real person sitting in front of the screen interacting with the application to see and check if every-thing works as intended. These testers are mostly non-technical people but have deep knowledge about the business cases in their respective modules / products.
 
Because this manual testing is obviously very inefficient, we developed a UI testing tool to help auto-mate the process. The testing tool basically has two jobs: Recording scripts and replaying the recorded scripts.
The way recording works is that the tester starts the testing tool, and the testing tool invokes the ap-plication under test. It does this in a way that lets the tester interact with the application as usual and records all the interactions of the tester into a script. The recording logic gets called from within the event listeners. We use derivations of the Vaadin components that add such listeners. 
 
Later on, that script can be replayed at a much higher speed than a human tester could do it. An arbi-trary number of such scripts can be compiled into a Jar and then be run sequentially on a testing envi-ronment such as virtual machines etc. 
 
During script replay, the commands from the test script call code that in turn does the actual work. We are trying to use testbench whenever possible if the API lets us. If not, we are forced to use some other ways like the WebElement API or custom JS. But these are options we would like to avoid.
The script replay logic also relies on the listeners mentioned above: Before a command from the script actually gets executed, it is added to a queue. It is now scheduled for execution. Inside the listener, the command gets dequeued if it is actually the one we expect. Else an error is thrown. This is how we verify that the command was actually executed, and the app responded as expected.
 
Therefore, it is important that the triggered events during script replay are exactly the same, in the exact same order as they were during script recording. Only then can we be sure that during replay no events are left out and no excess events are triggered. 
 
Now, coming back to the issue of this ticket. If there is a ContextMenu which is closed, we need to know by what means it was closed. Because if we just close it "somehow" during replay, we might set off the wrong events (compared to the ones that fired during recording). This may lead to different behavior of the app. It will then look like the test has failed when in fact we just triggered a wrong event during replay.

**Comment by rolfsmeds** [comment]
Thanks for the context, @haenni. The main thing I'm wondering is whether the _way_ in which the ContextMenu is "cancelled" (i.e. closed without triggering an action) is actually relevant: does it matter whether the user clicked outside or pressed Esc to close it, seeing as how both actions are handled by the component itself and considered equivalent from its point of view?

If not, it seems to me that we could simply add a boolean property that reveals whether the menu was "cancelled" or closed upon triggering an action. (And I would rather introduce a very simple API for what is, after all, a very niche requirement.)

**Comment by haenni** [comment]
@rolfsmeds Thank you for looking into it.

If there would be a boolean property on the event as you suggest then that would be sufficient to detect cancelling without action. 
We would however also need a mechanism to check whether the cancelling was done from the server side. Because if the cancelling was done from the server we need to prevent recording into the test script.

Actually, the only thing that prevents us from using openChangeListener in its current state is the fact that `com.vaadin.flow.component.contextmenu.ContextMenuBase#close` calls JS on the client. If the event triggered by that method would have `fromClient` == false, then that would also solve our problem and I think that would also be the cleanest solution.

**Comment by rolfsmeds** [comment]
Thanks @haenni, a `fromClient` property does seem reasonable to me.

**Comment by rolfsmeds** [comment]
Actually, I decided to replace this with a bug issue, since of course `isFromClient` should return a value that actually means something: https://github.com/vaadin/flow-components/issues/6449

(If anyone feels like this should be kept as a potential new API, let me know.)
